pkg/controller/secretgenerator/secretgenerator_controller.go

```diff
@@ -0,0 +1,139 @@
+package secretgenerator
+
+import (
+	"context"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// ReconcileSecretGenerator holds the controller configuration
+type ReconcileSecretGenerator struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// newReconciler initializes the ReconcileSecretGenerator object as
+// reconcile.Reconciler type
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileSecretGenerator{
+		client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+}
+
+// AddToManager creates a new controller and adds it to the manager. The manager
+// will set fields on the controller and start it when the manager is started.
+func AddToManager(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// add adds a new controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(
+		"secretgenerator-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch only new secrets with the corresponding annotation
+	predicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			object, _ := meta.Accessor(e.Object)
+			if _, found := object.GetAnnotations()[SECRET_NAME_ANNOTATION]; found {
+				return true
+			}
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
```

Comment:
Should we also watch for Secret updates?

Reply:
When the secret is mounted in a pod, we don't want to re-generate the random value if the secret is updated to avoid a drift between the data content and the data mounted in the deployment (deployments do not restart when secrets are updated). The random value will be only generated for new secrets.

Reply:
Okay, then this predicate looks good.